// ngx-slickjs/esm2015/lib/ngx-slickjs.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from "@angular/core";
import { SlickContainerDirective } from "./directives/slick-container.directive";
import { SlickItemDirective } from "./directives/slick-item.directive";
import { SLICK_LINKS, linkOptionsFactory } from "./tokens/links.token";
export class NgxSlickJsModule {
    /**
     * @param {?=} __0
     * @return {?}
     */
    static forRoot({ links = {} } = (/** @type {?} */ ({}))) {
        return {
            ngModule: NgxSlickJsModule,
            providers: [
                {
                    provide: SLICK_LINKS,
                    useValue: links,
                },
                {
                    provide: "slick-links",
                    useFactory: linkOptionsFactory,
                    deps: [SLICK_LINKS],
                },
            ],
        };
    }
}
NgxSlickJsModule.decorators = [
    { type: NgModule, args: [{
                declarations: [SlickContainerDirective, SlickItemDirective],
                exports: [SlickContainerDirective, SlickItemDirective],
            },] }
];
