// ngx-slickjs/bundles/ngx-slickjs.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('just-compare'), require('rxjs/operators'), require('rxjs')) :
    typeof define === 'function' && define.amd ? define('ngx-slickjs', ['exports', '@angular/core', 'just-compare', 'rxjs/operators', 'rxjs'], factory) :
    (global = global || self, factory(global['ngx-slickjs'] = {}, global.ng.core, global.compare, global.rxjs.operators, global.rxjs));
}(this, function (exports, core, compare, operators, rxjs) { 'use strict';

    compare = compare && compare.hasOwnProperty('default') ? compare['default'] : compare;

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (_) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    function __exportStar(m, exports) {
        for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
    }

    function __values(o) {
        var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
        if (m) return m.call(o);
        return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    };

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
        result.default = mod;
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var LazyLoadService = /** @class */ (function () {
        function LazyLoadService() {
            this._loadedLibraries = {};
        }
        /**
         * @param {?} url
         * @param {?} type
         * @return {?}
         */
        LazyLoadService.prototype.load = /**
         * @param {?} url
         * @param {?} type
         * @return {?}
         */
        function (url, type) {
            var _this = this;
            if (!url)
                return rxjs.of(null);
            /** @type {?} */
            var key = url.slice(url.lastIndexOf("/") + 1);
            if (this._loadedLibraries[key]) {
                return this._loadedLibraries[key].asObservable();
            }
            this._loadedLibraries[key] = new rxjs.ReplaySubject();
            /** @type {?} */
            var library = document.createElement(type === "script" ? "script" : "link");
            if (type === "script") {
                library.type = "text/javascript";
                ((/** @type {?} */ (library))).src = url;
            }
            else {
                library.type = "text/css";
                ((/** @type {?} */ (library))).rel = "stylesheet";
                ((/** @type {?} */ (library))).href = url;
            }
            library.onload = (/**
             * @return {?}
             */
            function () {
                _this._loadedLibraries[key].next();
                _this._loadedLibraries[key].complete();
            });
            document.body.appendChild(library);
            return this._loadedLibraries[key].asObservable();
        };
        LazyLoadService.decorators = [
            { type: core.Injectable, args: [{
                        providedIn: "root",
                    },] }
        ];
        /** @nocollapse */ LazyLoadService.ngInjectableDef = core.ɵɵdefineInjectable({ factory: function LazyLoadService_Factory() { return new LazyLoadService(); }, token: LazyLoadService, providedIn: "root" });
        return LazyLoadService;
    }());
    if (false) {
        /** @type {?} */
        LazyLoadService.prototype._loadedLibraries;
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var SlickContainerDirective = /** @class */ (function () {
        function SlickContainerDirective(elRef, zone, lazyLoadService, links) {
            this.elRef = elRef;
            this.zone = zone;
            this.lazyLoadService = lazyLoadService;
            this.links = links;
            this.config = (/** @type {?} */ ({}));
            this.afterChange = new core.EventEmitter();
            this.beforeChange = new core.EventEmitter();
            this.breakpoint = new core.EventEmitter();
            this.destroy = new core.EventEmitter();
            this.init = new core.EventEmitter();
            this.slides = [];
            this.initialize = false;
            this.slickCount = 0;
            this.destroy$ = new rxjs.Subject();
        }
        /**
         * @return {?}
         */
        SlickContainerDirective.prototype.ngAfterViewInit = /**
         * @return {?}
         */
        function () {
            var _this = this;
            this.lazyLoadService
                .load(this.links.jquery, 'script')
                .pipe(operators.map((/**
             * @return {?}
             */
            function () { return 'jQuery is loaded'; })), operators.filter((/**
             * @param {?} jquery
             * @return {?}
             */
            function (jquery) { return !!jquery; })), operators.switchMap((/**
             * @return {?}
             */
            function () {
                return rxjs.forkJoin(_this.lazyLoadService.load(_this.links.slickJs, 'script'), _this.lazyLoadService.load(_this.links.slickCss, 'style'), _this.lazyLoadService.load(_this.links.slickThemeCss, 'style'));
            })), operators.take(1))
                .subscribe((/**
             * @return {?}
             */
            function () {
                _this.initSlick();
            }));
        };
        /**
         * @return {?}
         */
        SlickContainerDirective.prototype.ngOnDestroy = /**
         * @return {?}
         */
        function () {
            this.unslick();
            this.slides = [];
            this.destroy$.next();
        };
        /**
         * @return {?}
         */
        SlickContainerDirective.prototype.initSlick = /**
         * @return {?}
         */
        function () {
            var _this = this;
            /** @type {?} */
            var that = this;
            this.zone.runOutsideAngular((/**
             * @return {?}
             */
            function () {
                $(_this.elRef.nativeElement)[0].innerHTML = '';
                _this.jQueryElement = $(_this.elRef.nativeElement);
                _this.jQueryElement.on('init', (/**
                 * @param {?} event
                 * @param {?} slick
                 * @return {?}
                 */
                function (event, slick) {
                    _this.zone.run((/**
                     * @return {?}
                     */
                    function () {
                        rxjs.timer(0).subscribe((/**
                         * @return {?}
                         */
                        function () {
                            _this.init.emit({ event: event, slick: slick });
                            if (_this.config && _this.config.initialSlide && _this.config.initialSlide <= _this.slides.length) {
                                _this.goTo(_this.config.initialSlide);
                            }
                        }));
                    }));
                }));
                _this.jQueryElement.slick(_this.config);
                _this.initialize = true;
                _this.listenMouseWheel();
                _this.jQueryElement.on('afterChange', (/**
                 * @param {?} event
                 * @param {?} slick
                 * @param {?} currentSlide
                 * @return {?}
                 */
                function (event, slick, currentSlide) {
                    that.zone.run((/**
                     * @return {?}
                     */
                    function () {
                        that.afterChange.emit({ event: event, slick: slick, currentSlide: currentSlide });
                    }));
                }));
                _this.jQueryElement.on('beforeChange', (/**
                 * @param {?} event
                 * @param {?} slick
                 * @param {?} currentSlide
                 * @param {?} nextSlide
                 * @return {?}
                 */
                function (event, slick, currentSlide, nextSlide) {
                    that.zone.run((/**
                     * @return {?}
                     */
                    function () {
                        _this.slickCount = slick.slideCount;
                        that.beforeChange.emit({ event: event, slick: slick, currentSlide: currentSlide, nextSlide: nextSlide });
                    }));
                }));
                _this.jQueryElement.on('breakpoint', (/**
                 * @param {?} event
                 * @param {?} slick
                 * @param {?} breakpoint
                 * @return {?}
                 */
                function (event, slick, breakpoint) {
                    that.zone.run((/**
                     * @return {?}
                     */
                    function () {
                        that.breakpoint.emit({ event: event, slick: slick, breakpoint: breakpoint });
                    }));
                }));
                _this.jQueryElement.on('destroy', (/**
                 * @param {?} event
                 * @param {?} slick
                 * @return {?}
                 */
                function (event, slick) {
                    that.zone.run((/**
                     * @return {?}
                     */
                    function () {
                        that.destroy.emit({ event: event, slick: slick });
                    }));
                }));
            }));
            this.syncSlides();
        };
        /**
         * @return {?}
         */
        SlickContainerDirective.prototype.syncSlides = /**
         * @return {?}
         */
        function () {
            var _this = this;
            if (this.slickCount >= this.slides.length)
                return;
            this.slides.forEach((/**
             * @param {?} slide
             * @return {?}
             */
            function (slide) {
                _this.zone.run((/**
                 * @return {?}
                 */
                function () {
                    _this.jQueryElement.slick('slickAdd', slide);
                }));
            }));
        };
        /**
         * @param {?} slide
         * @return {?}
         */
        SlickContainerDirective.prototype.addSlide = /**
         * @param {?} slide
         * @return {?}
         */
        function (slide) {
            var _this = this;
            this.slides = __spread(this.slides, [slide]);
            if (!this.initialize)
                return;
            this.zone.run((/**
             * @return {?}
             */
            function () {
                _this.jQueryElement.slick('slickAdd', slide);
            }));
        };
        /**
         * @param {?} slide
         * @return {?}
         */
        SlickContainerDirective.prototype.removeSlide = /**
         * @param {?} slide
         * @return {?}
         */
        function (slide) {
            var _this = this;
            this.slides = this.slides.filter((/**
             * @param {?} slide
             * @return {?}
             */
            function (slide) { return compare(slide, slide); }));
            if (!this.initialize)
                return;
            this.zone.run((/**
             * @return {?}
             */
            function () {
                _this.jQueryElement.slick('slickRemove', _this.slides.indexOf(slide));
            }));
        };
        /**
         * @param {?} index
         * @return {?}
         */
        SlickContainerDirective.prototype.goTo = /**
         * @param {?} index
         * @return {?}
         */
        function (index) {
            var _this = this;
            this.zone.run((/**
             * @return {?}
             */
            function () {
                _this.jQueryElement.slick('slickGoTo', index);
            }));
        };
        /**
         * @return {?}
         */
        SlickContainerDirective.prototype.next = /**
         * @return {?}
         */
        function () {
            var _this = this;
            this.zone.run((/**
             * @return {?}
             */
            function () {
                _this.jQueryElement.slick('slickNext');
            }));
        };
        /**
         * @return {?}
         */
        SlickContainerDirective.prototype.prev = /**
         * @return {?}
         */
        function () {
            var _this = this;
            this.zone.run((/**
             * @return {?}
             */
            function () {
                _this.jQueryElement.slick('slickPrev');
            }));
        };
        /**
         * @return {?}
         */
        SlickContainerDirective.prototype.pause = /**
         * @return {?}
         */
        function () {
            var _this = this;
            this.zone.run((/**
             * @return {?}
             */
            function () {
                _this.jQueryElement.slick('slickPause');
            }));
        };
        /**
         * @return {?}
         */
        SlickContainerDirective.prototype.play = /**
         * @return {?}
         */
        function () {
            var _this = this;
            this.zone.run((/**
             * @return {?}
             */
            function () {
                _this.jQueryElement.slick('slickPlay');
            }));
        };
        /**
         * @return {?}
         */
        SlickContainerDirective.prototype.unslick = /**
         * @return {?}
         */
        function () {
            var _this = this;
            if (this.jQueryElement) {
                this.zone.run((/**
                 * @return {?}
                 */
                function () {
                    _this.jQueryElement.slick('unslick');
                }));
            }
            this.initialize = false;
        };
        /**
         * @return {?}
         */
        SlickContainerDirective.prototype.listenMouseWheel = /**
         * @return {?}
         */
        function () {
            var _this = this;
            if (this.config.mouseWheelMove === false)
                return;
            rxjs.fromEvent(this.elRef.nativeElement, 'wheel')
                .pipe(operators.takeUntil(this.destroy$))
                .subscribe((/**
             * @param {?} event
             * @return {?}
             */
            function (event) {
                event.preventDefault();
                if (event.deltaX > 0 || event.deltaY < 0) {
                    _this.prev();
                }
                else if (event.deltaX < 0 || event.deltaY > 0) {
                    _this.next();
                }
            }));
        };
        SlickContainerDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: '[slickContainer]',
                        exportAs: 'slick',
                    },] }
        ];
        /** @nocollapse */
        SlickContainerDirective.ctorParameters = function () { return [
            { type: core.ElementRef },
            { type: core.NgZone },
            { type: LazyLoadService },
            { type: undefined, decorators: [{ type: core.Inject, args: ['slick-links',] }] }
        ]; };
        SlickContainerDirective.propDecorators = {
            config: [{ type: core.Input, args: ['slickConfig',] }],
            afterChange: [{ type: core.Output, args: ['slickAfterChange',] }],
            beforeChange: [{ type: core.Output, args: ['slickBeforeChange',] }],
            breakpoint: [{ type: core.Output, args: ['slickBreakpoint',] }],
            destroy: [{ type: core.Output, args: ['slickDestroy',] }],
            init: [{ type: core.Output, args: ['slickInit',] }]
        };
        return SlickContainerDirective;
    }());
    if (false) {
        /** @type {?} */
        SlickContainerDirective.prototype.config;
        /** @type {?} */
        SlickContainerDirective.prototype.afterChange;
        /** @type {?} */
        SlickContainerDirective.prototype.beforeChange;
        /** @type {?} */
        SlickContainerDirective.prototype.breakpoint;
        /** @type {?} */
        SlickContainerDirective.prototype.destroy;
        /** @type {?} */
        SlickContainerDirective.prototype.init;
        /** @type {?} */
        SlickContainerDirective.prototype.slides;
        /** @type {?} */
        SlickContainerDirective.prototype.jQueryElement;
        /** @type {?} */
        SlickContainerDirective.prototype.initialize;
        /** @type {?} */
        SlickContainerDirective.prototype.slickCount;
        /**
         * @type {?}
         * @private
         */
        SlickContainerDirective.prototype.destroy$;
        /**
         * @type {?}
         * @private
         */
        SlickContainerDirective.prototype.elRef;
        /**
         * @type {?}
         * @private
         */
        SlickContainerDirective.prototype.zone;
        /**
         * @type {?}
         * @private
         */
        SlickContainerDirective.prototype.lazyLoadService;
        /**
         * @type {?}
         * @private
         */
        SlickContainerDirective.prototype.links;
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var SlickItemDirective = /** @class */ (function () {
        function SlickItemDirective(elRef, slickRef) {
            this.elRef = elRef;
            this.slickRef = slickRef;
        }
        /**
         * @return {?}
         */
        SlickItemDirective.prototype.ngAfterViewInit = /**
         * @return {?}
         */
        function () {
            this.slickRef.addSlide(this.elRef.nativeElement);
        };
        /**
         * @return {?}
         */
        SlickItemDirective.prototype.ngOnDestroy = /**
         * @return {?}
         */
        function () {
            this.slickRef.removeSlide(this.elRef.nativeElement);
        };
        SlickItemDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: "[slickItem]"
                    },] }
        ];
        /** @nocollapse */
        SlickItemDirective.ctorParameters = function () { return [
            { type: core.ElementRef },
            { type: SlickContainerDirective, decorators: [{ type: core.Host }] }
        ]; };
        return SlickItemDirective;
    }());
    if (false) {
        /** @type {?} */
        SlickItemDirective.prototype.elRef;
        /**
         * @type {?}
         * @private
         */
        SlickItemDirective.prototype.slickRef;
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * @param {?} links
     * @return {?}
     */
    function linkOptionsFactory(links) {
        return (/** @type {?} */ (__assign({ jquery: "https://code.jquery.com/jquery-3.4.0.min.js", slickJs: "https://cdn.jsdelivr.net/npm/slick-carousel@1.8.1/slick/slick.min.js", slickCss: "https://cdn.jsdelivr.net/npm/slick-carousel@1.8.1/slick/slick.css", slickThemeCss: "https://cdn.jsdelivr.net/npm/slick-carousel@1.8.1/slick/slick-theme.css" }, links)));
    }
    /** @type {?} */
    var SLICK_LINKS = new core.InjectionToken("slick-links");

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var NgxSlickJsModule = /** @class */ (function () {
        function NgxSlickJsModule() {
        }
        /**
         * @param {?=} __0
         * @return {?}
         */
        NgxSlickJsModule.forRoot = /**
         * @param {?=} __0
         * @return {?}
         */
        function (_a) {
            var _b = (_a === void 0 ? (/** @type {?} */ ({})) : _a).links, links = _b === void 0 ? {} : _b;
            return {
                ngModule: NgxSlickJsModule,
                providers: [
                    {
                        provide: SLICK_LINKS,
                        useValue: links,
                    },
                    {
                        provide: "slick-links",
                        useFactory: linkOptionsFactory,
                        deps: [SLICK_LINKS],
                    },
                ],
            };
        };
        NgxSlickJsModule.decorators = [
            { type: core.NgModule, args: [{
                        declarations: [SlickContainerDirective, SlickItemDirective],
                        exports: [SlickContainerDirective, SlickItemDirective],
                    },] }
        ];
        return NgxSlickJsModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var Options;
    (function (Options) {
        /**
         * @record
         */
        function Root() { }
        Options.Root = Root;
        if (false) {
            /** @type {?} */
            Root.prototype.links;
        }
        /**
         * @record
         */
        function Links() { }
        Options.Links = Links;
        if (false) {
            /** @type {?|undefined} */
            Links.prototype.jquery;
            /** @type {?|undefined} */
            Links.prototype.slickJs;
            /** @type {?|undefined} */
            Links.prototype.slickCss;
            /** @type {?|undefined} */
            Links.prototype.slickThemeCss;
        }
    })(Options || (Options = {}));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var Slick;
    (function (Slick) {
        /**
         * @record
         */
        function Config() { }
        Slick.Config = Config;
        if (false) {
            /** @type {?|undefined} */
            Config.prototype.accessibility;
            /** @type {?|undefined} */
            Config.prototype.adaptiveHeight;
            /** @type {?|undefined} */
            Config.prototype.autoplay;
            /** @type {?|undefined} */
            Config.prototype.autoplaySpeed;
            /** @type {?|undefined} */
            Config.prototype.arrows;
            /** @type {?|undefined} */
            Config.prototype.asNavFor;
            /** @type {?|undefined} */
            Config.prototype.appendArrows;
            /** @type {?|undefined} */
            Config.prototype.appendDots;
            /** @type {?|undefined} */
            Config.prototype.prevArrow;
            /** @type {?|undefined} */
            Config.prototype.nextArrow;
            /** @type {?|undefined} */
            Config.prototype.centerMode;
            /** @type {?|undefined} */
            Config.prototype.centerPadding;
            /** @type {?|undefined} */
            Config.prototype.cssEase;
            /** @type {?|undefined} */
            Config.prototype.dots;
            /** @type {?|undefined} */
            Config.prototype.dotsClass;
            /** @type {?|undefined} */
            Config.prototype.draggable;
            /** @type {?|undefined} */
            Config.prototype.fade;
            /** @type {?|undefined} */
            Config.prototype.focusOnSelect;
            /** @type {?|undefined} */
            Config.prototype.easing;
            /** @type {?|undefined} */
            Config.prototype.edgeFriction;
            /** @type {?|undefined} */
            Config.prototype.infinite;
            /** @type {?|undefined} */
            Config.prototype.initialSlide;
            /** @type {?|undefined} */
            Config.prototype.lazyLoad;
            /** @type {?|undefined} */
            Config.prototype.mobileFirst;
            /** @type {?|undefined} */
            Config.prototype.pauseOnFocus;
            /** @type {?|undefined} */
            Config.prototype.pauseOnDotsHover;
            /** @type {?|undefined} */
            Config.prototype.respondTo;
            /** @type {?|undefined} */
            Config.prototype.responsive;
            /** @type {?|undefined} */
            Config.prototype.rows;
            /** @type {?|undefined} */
            Config.prototype.slidesPerRow;
            /** @type {?|undefined} */
            Config.prototype.slidesToShow;
            /** @type {?|undefined} */
            Config.prototype.slidesToScroll;
            /** @type {?|undefined} */
            Config.prototype.speed;
            /** @type {?|undefined} */
            Config.prototype.swipe;
            /** @type {?|undefined} */
            Config.prototype.swipeToSlide;
            /** @type {?|undefined} */
            Config.prototype.touchMove;
            /** @type {?|undefined} */
            Config.prototype.mouseWheelMove;
            /** @type {?|undefined} */
            Config.prototype.touchThreshold;
            /** @type {?|undefined} */
            Config.prototype.useCSS;
            /** @type {?|undefined} */
            Config.prototype.useTransform;
            /** @type {?|undefined} */
            Config.prototype.variableWidth;
            /** @type {?|undefined} */
            Config.prototype.vertical;
            /** @type {?|undefined} */
            Config.prototype.verticalSwiping;
            /** @type {?|undefined} */
            Config.prototype.rti;
            /** @type {?|undefined} */
            Config.prototype.waitForAnimate;
            /** @type {?|undefined} */
            Config.prototype.zIndex;
        }
        /**
         * @record
         */
        function ConfigResponsive() { }
        Slick.ConfigResponsive = ConfigResponsive;
        if (false) {
            /** @type {?} */
            ConfigResponsive.prototype.breakpoint;
            /** @type {?|undefined} */
            ConfigResponsive.prototype.settings;
        }
    })(Slick || (Slick = {}));

    exports.LazyLoadService = LazyLoadService;
    exports.NgxSlickJsModule = NgxSlickJsModule;
    exports.SLICK_LINKS = SLICK_LINKS;
    exports.SlickContainerDirective = SlickContainerDirective;
    exports.SlickItemDirective = SlickItemDirective;
    exports.linkOptionsFactory = linkOptionsFactory;
    exports.ɵa = SlickContainerDirective;
    exports.ɵb = LazyLoadService;
    exports.ɵd = SlickItemDirective;
    exports.ɵe = linkOptionsFactory;
    exports.ɵf = SLICK_LINKS;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
